packages/babel-core/src/transformation/internal-plugins/shadow-functions.js

```diff
@@ -92,7 +92,10 @@ function remap(path, key) {
 
   fnPath.setData(key, id);
 
-  if (key === "this" && fnPath.isMethod({kind: "constructor"})) {
+  let classPath = fnPath.findParent((p) => p.isProgram() || p.isClassDeclaration());
+  let hasSuperClass = !!(classPath && classPath.node && classPath.node.superClass);
+
+  if (key === "this" && fnPath.isMethod({kind: "constructor"}) && hasSuperClass) {
```

Comment:
Ugh, good call, I totally missed this case. This looks fine to me, but can we remove the `p.isProgram()` case? Since this is checking `!!classPath` we should be fine without it.